In [20]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine

In [2]:
engine = create_engine("postgres+psycopg2://postgres:postgres@localhost:5432/Lahman Baseball")

In [3]:
batting_df = pd.read_sql("SELECT * FROM batting;", con=engine)
schools_df = pd.read_sql("SELECT * FROM schools;", con=engine)
college_playing_df = pd.read_sql("SELECT * FROM collegeplaying;", con=engine)
people_df = pd.read_sql("SELECT * FROM people;", con=engine)
salaries_df = pd.read_sql("SELECT * FROM salaries;", con=engine)
fielding_df = pd.read_sql("SELECT * FROM fielding;", con=engine)
teams_df = pd.read_sql("SELECT * FROM teams;", con=engine)
teams_franchises_df = pd.read_sql("SELECT * FROM teamsfranchises;", con=engine)
managers_df = pd.read_sql("SELECT * FROM managers;", con=engine)
awards_managers_df = pd.read_sql("SELECT * FROM awardsmanagers;", con=engine)
appearances_df = pd.read_sql("SELECT * FROM appearances;", con=engine)
awards_players_df = pd.read_sql("SELECT * FROM awardsplayers;", con=engine)
halloffame_df = pd.read_sql("SELECT * FROM halloffame;", con=engine)
pitching_df = pd.read_sql("SELECT * FROM pitching;", con=engine)

#### 1 a. Find all players in the database who played at Vanderbilt University.
#### b. Create a list showing each player’s first and last names as well as the total salary they earned in the major leagues.
#### c. Sort this list in descending order by the total salary earned.
#### d. Which Vanderbilt player earned the most money in the majors?

In [89]:
vandy_df = pd.merge(schools_df, college_playing_df, on='schoolid')
vandy_df = pd.merge(vandy_df, people_df, on='playerid')
vandy_df = pd.merge(vandy_df, salaries_df, on='playerid')

In [95]:
vandy_sub = vandy_df[['schoolname', 'namefirst', 'namelast', 'salary', 'yearid_y']]
vandy_sub = vandy_sub[vandy_sub['schoolname'] == 'Vanderbilt University'].reset_index(drop=True)

In [96]:
unique_vandy = vandy_sub.sort_values('salary').drop_duplicates(subset=['namefirst', 'namelast', 'yearid_y'], keep='last').reset_index(drop=True)

In [98]:
vandy_sals = unique_vandy.groupby(['namefirst', 'namelast'], as_index=False)['salary'].sum()
vandy_sals.sort_values('salary', ascending=False).reset_index(drop=True)

namefirst    namelast      salary
0      David       Price  81851296.0
1      Pedro     Alvarez  20681704.0
2       Mark       Prior  12800000.0
3      Scott   Sanderson  10500000.0
4       Mike       Minor   6837500.0
5       Joey        Cora   5622500.0
6       Ryan    Flaherty   4061000.0
7       Josh        Paul   2640000.0
8       Mike      Baxter   2094418.0
9      Sonny        Gray   1542500.0
10    Jensen       Lewis   1234000.0
11      Matt        Kata   1060000.0
12      Nick  Christiani    500000.0
13    Jeremy      Sowers    384800.0
14    Scotti     Madison    135000.0

David Price has earned the most among players who attended Vanderbilt at $81,851,296

#### 2. Using the fielding table, group players into three groups based on their position: label players with position OF as "Outfield", those with position "SS", "1B", "2B", and "3B" as "Infield", and those with position "P" or "C" as "Battery". Determine the number of putouts made by each of these three groups in 2016.

In [100]:
fielding_df['position_group'] = 'Infield'
fielding_df.loc[fielding_df['pos'] == 'OF', 'position_group'] = 'Outfield'
fielding_df.loc[fielding_df['pos'] == 'P', 'position_group'] = 'Battery'
fielding_df.loc[fielding_df['pos'] == 'C', 'position_group'] = 'Battery'

In [108]:
fielding_2016 = fielding_df[fielding_df['yearid'] == 2016].reset_index(drop=True)
fielding_2016.groupby('position_group')['po'].sum()

position_group
Battery     41424
Infield     58934
Outfield    29560
Name: po, dtype: int64

Infield: 58,934
Battery: 41,424
Outfield: 29,560

#### 3. a. From 1970 – 2016, what is the largest number of wins for a team that did not win the world series?
#### b. What is the smallest number of wins for a team that did win the world series?
#### c. Doing this will probably result in an unusually small number of wins for a world series champion – determine why this is the case.
#### d. Then redo your query, excluding the problem year.
#### e. How often from 1970 – 2016 was it the case that a team with the most wins also won the world series?
#### f. What percentage of the time?

In [12]:
wins_base_df = pd.merge(teams_df, teams_franchises_df, on='franchid', how='left')
wins_ab_df = wins_base_df[['yearid','franchname','g','w','wswin']]
wins_ab_df = wins_ab_df[wins_ab_df['yearid'] > 1969].reset_index(drop=True)
wins_ab_df = wins_ab_df[wins_ab_df['yearid'] != 1994].reset_index(drop=True)
wins_a_df = wins_ab_df[wins_ab_df['wswin'] == 'N']
wins_b_df = wins_ab_df[wins_ab_df['wswin'] == 'Y']

yearid lgid teamid franchid divid  rank    g  ghome   w   l  ...  \
2830    2016   NL    SLN      STL     C     2  162   81.0  86  76  ...   
2831    2016   AL    TBA      TBD     E     5  162   81.0  68  94  ...   
2832    2016   AL    TEX      TEX     W     1  162   81.0  95  67  ...   
2833    2016   AL    TOR      TOR     E     2  162   81.0  89  73  ...   
2834    2016   NL    WAS      WSN     E     1  162   81.0  95  67  ...   

                               park attendance  bpf  ppf  teamidbr  \
2830              Busch Stadium III  3444490.0  100   99       STL   
2831                Tropicana Field  1286163.0   93   94       TBR   
2832  Rangers Ballpark in Arlington  2710402.0  106  105       TEX   
2833                  Rogers Centre  3392099.0  111  110       TOR   
2834                 Nationals Park  2481938.0  100   98       WSN   

      teamidlahman45  teamidretro            franchname  active  naassoc  
2830             SLN          SLN   St. Louis Cardinals       Y     None  
2831             TBA          TBA        Tampa Bay Rays       Y     None  
2832             TEX          TEX         Texas Rangers       Y     None  
2833             TOR          TOR     Toronto Blue Jays       Y     None  
2834             MON          WAS  Washington Nationals       Y     None  

[5 rows x 51 columns]

In [40]:
wins_a_df = wins_a_df.reset_index(drop=True)
wins_a_df.iloc[wins_a_df['w'].idxmax()]

yearid                    2001
franchname    Seattle Mariners
g                          162
w                          116
wswin                        N
Name: 779, dtype: object

The 2001 Seattle Mariners won 116 games and didn't win the World Series

In [15]:
wins_b_df = wins_b_df.reset_index(drop=True)
wins_b_df.iloc[wins_b_df['w'].idxmin()]

yearid                       1981
franchname    Los Angeles Dodgers
g                             110
w                              63
wswin                           Y
Name: 11, dtype: object

The 1981 Los Angeles Dodgers won 63 games en route to winning the World Series

The Dodgers only played 110 games in 1981, as a substantial portion of the season was lost to the 1981 MLB strike.

In [18]:
wins_d_df = wins_b_df[wins_b_df['yearid'] != 1981]
wins_d_df = wins_d_df.reset_index(drop=True)
wins_d_df.iloc[wins_d_df['w'].idxmin()]

yearid                       2006
franchname    St. Louis Cardinals
g                             161
w                              83
wswin                           Y
Name: 34, dtype: object

The 2006 St. Louis Cardinals won 83 games and the World Series in a standard 162 game season.

In [53]:
winfilter = wins_ab_df.groupby('yearid')['w'].transform('max') == wins_ab_df['w']
wins_e_df = wins_ab_df.loc[winfilter].reset_index(drop=True)
wins_e_df['wswin'].value_counts()

N    40
Y    12
Name: wswin, dtype: int64

In [57]:
round(12/46,3)

0.261

From 1970-2016, 12 teams had the most wins and won the World Series. Since there was no World Series in 1994, there are 46 seasons where there was the opportunity for the wins leader to win the World Series, so it happened 26.1% of the time.

#### 4. Which managers have won the TSN Manager of the Year award in both the National League (NL) and the American League (AL)? Give their full name and the teams that they were managing when they won the award.

In [138]:
tsnmoy_df = pd.merge(awards_managers_df, people_df, on='playerid')
tsnmoy_df = pd.merge(tsnmoy_df, managers_df, on=['playerid','yearid'])
tsnmoy_df = pd.merge(tsnmoy_df, teams_df, on=['teamid','yearid'])
tsnmoy_sub = tsnmoy_df[['namefirst', 'namelast', 'yearid', 'awardid', 'lgid_x', 'name']]
tsnmoy_sub = tsnmoy_sub[tsnmoy_sub['awardid'] == 'TSN Manager of the Year'].reset_index(drop=True)
tsnmoy_sub = tsnmoy_sub[tsnmoy_sub['lgid_x'].isin(['AL','NL'])].reset_index(drop=True)
tsnmoy_sub['fullname']= tsnmoy_sub['namefirst'] + ' ' + tsnmoy_sub['namelast']
tsnmoy_sub = tsnmoy_sub.drop(columns=['namefirst', 'namelast'])
tsnmoy_sub.columns = ['yearid','awardid','lgid','team_name','full_name']
tsnmoy_sub = tsnmoy_sub[['full_name','yearid','lgid','team_name','awardid']]
tsnmoy_sub.head(5)

full_name  yearid lgid          team_name                  awardid
0     Tony LaRussa    1988   AL  Oakland Athletics  TSN Manager of the Year
1     Tony LaRussa    1992   AL  Oakland Athletics  TSN Manager of the Year
2  Sparky Anderson    1987   AL     Detroit Tigers  TSN Manager of the Year
3        Bobby Cox    1991   NL     Atlanta Braves  TSN Manager of the Year
4        Bobby Cox    2004   NL     Atlanta Braves  TSN Manager of the Year

In [141]:
tsnmoy_sub.groupby(['full_name','lgid']).size()

full_name        lgid
Bob Melvin       NL      1
Bobby Cox        NL      7
Bruce Bochy      NL      2
Buck Rodgers     NL      1
Buck Showalter   AL      3
Buddy Black      NL      1
Clint Hurdle     NL      1
Davey Johnson    AL      1
                 NL      1
Don Baylor       NL      1
Don Zimmer       NL      1
Dusty Baker      NL      2
Eric Wedge       AL      1
Felipe Alou      NL      1
Frank Robinson   AL      1
Fredi Gonzalez   NL      1
Hal Lanier       NL      1
Jeff Torborg     AL      1
Jerry Manuel     AL      1
Jim Leyland      AL      1
                 NL      3
Jim Tracy        NL      1
Jimy Williams    AL      1
Joe Girardi      NL      1
Joe Maddon       AL      2
Joe Torre        AL      1
John Farrell     AL      1
John McNamara    AL      1
Johnny Oates     AL      2
Kirk Gibson      NL      1
Larry Bowa       NL      1
Lou Piniella     AL      1
Matt Williams    NL      1
Mike Hargrove    AL      1
Mike Scioscia    AL      2
Ozzie Guillen    AL      1
Paul M

In [142]:
tsnmoy_bash = tsnmoy_sub[tsnmoy_sub['full_name'].isin(['Davey Johnson','Jim Leyland'])].reset_index(drop=True)
tsnmoy_bash

full_name  yearid lgid             team_name                  awardid
0    Jim Leyland    1990   NL    Pittsburgh Pirates  TSN Manager of the Year
1    Jim Leyland    1992   NL    Pittsburgh Pirates  TSN Manager of the Year
2    Jim Leyland    2006   AL        Detroit Tigers  TSN Manager of the Year
3    Jim Leyland    1988   NL    Pittsburgh Pirates  TSN Manager of the Year
4  Davey Johnson    1997   AL     Baltimore Orioles  TSN Manager of the Year
5  Davey Johnson    2012   NL  Washington Nationals  TSN Manager of the Year